docker/ssl/authorization.js

```diff
@@ -0,0 +1,166 @@
+
+function authenticate(s) {
+    if (!s.variables.ssl_client_s_dn || !s.variables.ssl_client_s_dn.length) {
+        s.deny();
+        return
+    }
+
+    s.on('upload', function (data) {
+        while (data == '') {
+            return s.AGAIN
+        }
+
+        if (clientKey === '') {
+            clientKey = parseCert(s.variables.ssl_client_s_dn, 'CN');
+        }
+
+        var pass = parsePackage(s, data);
+
+        if (!clientKey.length || pass !== clientKey) {
+            s.error('Cert CN (' + clientKey + ') does not match client ID');
+            s.off('upload')
+            s.deny();
+        }
+
+        s.off('upload');
+        s.allow();
+    })
+}
+
+function parsePackage(s, data) {
+    // An explanation of MQTT packet structure can be found here:
+    // https://public.dhe.ibm.com/software/dw/webservices/ws-mqtt/mqtt-v3r1.html#msg-format. 
+
+    // CONNECT message is explained here:
+    // https://public.dhe.ibm.com/software/dw/webservices/ws-mqtt/mqtt-v3r1.html#connect.
+
+    /*
+        0               1               2               3
+	    7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0 7 6 5 4 3 2 1 0
+	    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+	    | TYPE | RSRVD | REMAINING LEN |      PROTOCOL NAME LEN       |
+	    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+	    |                        PROTOCOL NAME                        |    
+	    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-|
+	    |    VERSION   |     FLAGS     |          KEEP ALIVE          | 
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-|
+	    |                     Payload (if any) ...                    |
+	    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+       
+        First byte with remaining length represents fixed header.
+        Remaining Length is the length of the variable header (10 bytes) plus the length of the Payload.
+        It is encoded in the manner described here:
+        http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/errata01/os/mqtt-v3.1.1-errata01-os-complete.html#_Toc442180836.
+        
+        Connect flags byte looks like this:   
+        |       7       |       6       |       5     |   4  3   |     2     |     1        |     0     |
+        | Username Flag | Password Flag | Will Retain | Will QoS | Will Flag | Clean Session | Reserved |
+
+        The payload is determined by the flags and comes in this order:
+            1. Client ID (2 bytes length + ID value)
+            2. Will Topic (2 bytes length + Will Topic value) if Will Flag is 1.
+            3. Will Message (2 bytes length + Will Message value) if Will Flag is 1.
+            4. User Name (2 bytes length + User Name value) if User Name Flag is 1.
+            5. Password (2 bytes length + Password value) if Password Flag is 1.
+        
+        This method extracts Password field.
+	*/
+
+    var packet_type_flags_byte = data.codePointAt(0);
```

Comment:
Llooks like you started parsing packet flags fields without even checking if the message type is `CONNECT`. If it is not `CONNECT`, we should exit right after this check of a first byte.